MoreLinq/Memoize.cs

```diff
@@ -0,0 +1,144 @@
+        /// be shared between threads. Each thread can call its
+        /// <see cref="IEnumerable{T}.GetEnumerator"/> to acquire an iterator
+        /// but the same iterator should not be used simultanesouly from
+        /// multiple threads. The sequence supplied in <paramref name="source"/>
+        /// is not expected to be thread-safe but it is required to be
+        /// thread-agnostic because different threads (though never
+        /// simultaneously) may iterate over the sequence.
+        /// </remarks>
+
+        public static IEnumerable<T> Memoize<T>(this IEnumerable<T> source)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+
+            if (source is ICollection<T>)
+            {
+                return source;
+            }
+
+            return (source as MemoizedEnumerable<T>) ?? new MemoizedEnumerable<T>(source);
+        }
+    }
+
+    internal class MemoizedEnumerable<T> : IEnumerable<T>, IDisposable
+    {
+        private IList<T> cache;
+        private readonly object locker;
+        private readonly IEnumerable<T> source;
+        private IEnumerator<T> sourceEnumerator;
+
+        public MemoizedEnumerable(IEnumerable<T> sequence)
+        {
+            if (sequence == null) throw new ArgumentNullException(nameof(sequence));
+
+            source = sequence;
+            locker = new object();
+        }
+
+        public IEnumerator<T> GetEnumerator()
+        {
+            if (cache == null)
+            {
+                lock (locker)
+                {
+                    if (cache == null)
+                    {
+                        var cache = new List<T>(); // for exception safety, allocate then...
+                        sourceEnumerator = source.GetEnumerator(); // (because this can fail)
+                        this.cache = cache; // ...commit to state
+                    }
+                }
+            }
+
+            var index = 0;
+
+            while (true)
+            {
+                bool hasValue;
+                lock (locker)
+                {
+                    if (index < cache.Count)
+                    {
+                        hasValue = true;
+                    }
+                    else if (sourceEnumerator == null)
+                    {
+                        hasValue = false;
+                    }
+                    else if (sourceEnumerator.MoveNext()) // TODO exception safety?
+                    {
+                        hasValue = true;
+                        cache.Add(sourceEnumerator.Current);
+                    }
+                    else
+                    {
+                        hasValue = false;
+                        sourceEnumerator.Dispose();
+                        sourceEnumerator = null;
+                    }
+                }
+
+                if (hasValue)
+                    yield return cache[index];
```

Comment:
This accesses the cache outside the lock. Maybe it's best to have another temporary variable `T current` with same scope as `hasValue`, set that in the lock section above, and yield it here.